static/src/javascripts/projects/common/modules/identity/global/sign-in-engagement-banner.js

```diff
@@ -0,0 +1,209 @@
+        userPrefs.set(sessionStartedAtKey, Date.now());
+        userPrefs.set(sessionVisitsKey, 0);
+    }
+    const sessionVisits: number = userPrefs.get(sessionVisitsKey) || 0;
+    userPrefs.set(sessionVisitsKey, sessionVisits + 1);
+};
+
+const features: Feature[] = [
+    {
+        icon: iconPhone.markup,
+        mainCopy: 'A consistent experience',
+        subCopy: 'across all of your devices',
+    },
+    {
+        icon: iconComment.markup,
+        mainCopy: 'Join the conversation',
+        subCopy: 'and comment on articles',
+    },
+    {
+        icon: iconEmail.markup,
+        mainCopy: 'Get closer to the journalism',
+        subCopy: 'by subscribing to editorial emails',
+    },
+];
+
+const tpl: Template = {
+    headerMain: ['Enjoy even', 'more from', 'The Guardian'],
+    headerSub: ['Please sign in or register to manage your preferences'],
+    signInCta: 'Sign in',
+    registerCta: 'Register',
+    advantagesCta: 'Why sign in to The Guardian?',
+    closeButton: 'Continue without signing in',
+    features,
+    links,
+};
+
+/* Must have visited 4 articles */
+const hasReadOver4Articles = (): boolean =>
+    (local.get('gu.alreadyVisited') || 0) >= 4;
+
+/* Must be not already signed in */
+const isNotSignedIn = (): boolean => getCookie(signedInCookie) === null;
+
+/* Must be shown only 4 times total */
+const hasSeenBannerLessThanFourTimesTotal = (): boolean =>
+    (userPrefs.get(lifeTimeViewsKey) || 0) < 4;
+
+/* Must be shown only once every 2 days */
+const hasSeenBannerOnceInLastTwoDays = (): boolean =>
+    Date.now() - (userPrefs.get(lastSeenAtKey) || Date.now() - monthInMs) >
+    dayInMs * 2;
+
+/* Must be shown on the second session pageview */
+const isSecondSessionPageview = (): boolean =>
+    (userPrefs.get(sessionVisitsKey) || 0) >= 2;
+
+/* Must have visited between 1 month & 24 hours ago */
+const isRecurringVisitor = (): boolean => {
+    const ga: ?string = getCookie('_ga');
+    if (!ga) return false;
+    const date: number = parseInt(ga.split('.').pop(), 10) * 1000;
+    if (!date || Number.isNaN(date)) return false;
+    return Date.now() - date > dayInMs && Date.now() - date < monthInMs;
+};
+
+const bannerDoesNotCollide = (): Promise<boolean> =>
+    new Promise(show => {
+        setTimeout(() => {
+            show(true);
+        }, 1000);
+
+        mediator.on('modules:onwards:breaking-news:ready', breakingShown => {
+            if (!breakingShown) {
+                show(true);
+            } else {
+                show(false);
+            }
+        });
+        mediator.on('membership-message:display', () => {
+            show(false);
+        });
+    });
+
+const hide = (msg: Message) => {
+    userPrefs.set(
+        lifeTimeClosesKey,
+        (userPrefs.get(lifeTimeClosesKey) || 0) + 1
+    );
+    msg.hide();
+};
+
+const canShow = (): Promise<boolean> => {
+    /* sorry for the side-effects */
+    recordSessionVisit();
```

Comment:
Can't do much about the side effect, but does it belong in this function? Maybe it make more sense at the module scope?

Reply:
there's a third way out: Is session visits (defined by visits in the last 30 minutes) something we want globally?

Reply:
Is this the only module that's currently recording this? I'd say wait and see if any other modules need to do this before extracting it.

Reply:
it is AFAIK yes